### PR TITLE
RDoc-2242-Fix the C# article + Add JS indexes

### PR DESCRIPTION
**Pull request by Danielle9897**
For issue: 
https://issues.hibernatingrhinos.com/issue/RDoc-2242/Fix-createField-sample-and-check-the-rest-of-the-Indexes-Dynamic-Fields-article

### CODE REVIEW

**Comment by Danielle9897** [comment]
**Still todo:**
Java + Node.js pages

**Comment by Danielle9897** [comment]
for @ml054 (and for anyone else reviewing):
pls review the latest files, **the 5.2 versions**

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java [hunk 1]
```diff
@@ -1,87 +1,192 @@
...
+      
+        //region dynamic_fields_10
+        public class Product_4
+        {
+            private String id;
+            private String name;
+            
+            // For each element in this list, the VALUE of property 'propName' will be dynamically indexed
+            // e.g. Color, Width, Length (in ex. below) will become dynamic-index-fields
+            private ArrayList<Attribute> attributes;
```
Comment (ml054) [inline Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:120]: ArrayList -> List

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java [hunk 2]
```diff
@@ -1,87 +1,192 @@
...
     public DynamicFields() {
         try (IDocumentStore store = new DocumentStore()) {
             try (IDocumentSession session = store.openSession()) {
-                //region dynamic_fields_4
-                List<Product> results = session
-                    .query(Product.class, Products_ByAttribute.class)
-                    .whereEquals("color", "red")
+                //region dynamic_fields_3
+                List<Product_1> matchingDocuments = session
+                    .advanced()
```
Comment (ml054) [inline Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:155]: session.advanved.docuemtnQuery -> session.query  (this is alias in java)

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java [hunk 3]
```diff
@@ -1,87 +1,192 @@
...
-    }
-
-    //region syntax
-    object CreateField(string name, object value);
 
-    object CreateField(string name, object value, bool stored, bool analyzed);
+            try (IDocumentSession session = store.openSession()) {
+                //region dynamic_fields_6
+                List<Product_2> matchingDocuments = session
+                    .advanced()
```
Comment (ml054) [inline Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:165]: same here and below

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java [hunk 4]
```diff
@@ -1,87 +1,192 @@
...
+            }
 
-    object CreateField(string name, object value, CreateFieldOptions options);
-    //endregion
+            try (IDocumentSession session = store.openSession()) {
+                //region dynamic_fields_9
+                List<Product_3> matchingDocuments = session
+                    .advanced()
+                    .documentQuery(Product_3.class, Products_ByProductType.class)
+                    .whereEquals("Electronics", 23)
```
Comment (ml054) [inline Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:177]: here we have same issue like in demo - upper vs lower case - are you sure it is ok?

Comment (Danielle9897) [reply Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:177]: well, in this case, 'Electronics' is not the field property name, it is its **value** (the persons's job/title)
so I thought it was ok uppercase

all props names are lower

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java (hunk 1 above)
Comment (Danielle9897) [reply Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:120]: done

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java (hunk 2 above)
Comment (Danielle9897) [reply Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:155]: right... done

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java (hunk 3 above)
Comment (Danielle9897) [reply Documentation/5.2/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java:165]: done

**Comment by Danielle9897** [comment]
@ppekrol  who should review the C# page ?

**Review by ml054 — APPROVED** [review]

**Comment by Danielle9897** [comment]
@ppekrol  - as suggested above - review may focus on 5.2 - thanks

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 5]
```diff
@@ -3,101 +3,281 @@
...
+            object CreateField(string fieldName, object fieldValue, bool stored, bool analyzed);
 
-            object CreateField(string name, object value, CreateFieldOptions options);
+            object CreateField(string fieldName, object fieldValue, CreateFieldOptions options);
             #endregion
         }
 
         #region dynamic_fields_1
-        public class Product
+        public class Product_1
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:22]: Product_1?

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 6]
```diff
@@ -3,101 +3,281 @@
...
+                // Using '_' is just a convention. Any other string can be used instead of '_'
+                // The actual field name generated is defined by method CreateField below
+                public object _ { get; set; }
+            }
 
-        public class Attribute
+            public Products_ByAttributeKey()
+            {
+                Map = products => from p in products
+                    select new Products_ByAttributeKey.IndexEntry
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:46]: better to use anonymous here maybe?

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 7]
```diff
@@ -3,101 +3,281 @@
...
+            // The KEYS under the Attributes object will be dynamically indexed 
+            // Fields added to this object after index creation time will also get indexed
+            public Dictionary<string, object> Attributes { get; set; }
         }
+        #endregion
+        
+        #region dynamic_fields_2
+        public class Products_ByAttributeKey : AbstractIndexCreationTask<Product_1>
+        {
+            public class IndexEntry
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:35]: not needed if anonymous is used?

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 8]
```diff
@@ -3,101 +3,281 @@
...
+                                { indexing: 'Search', storage: true, termVector: null }))
+                        };
+                    })"
+                };
+            }
         }
         #endregion
+        
+        #region dynamic_fields_4
+        public class Product_2
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:78]: this is a problem, maybe separate namespaces?

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 9]
```diff
@@ -3,101 +3,281 @@
...
+            {
+                public object _ { get; set; }
+            }
+
+            public Products_ByAnyField_JS()
+            {
+                // this will index EVERY FIELD under the top level of the document
+                Maps = new HashSet<string>
+                {
+                    @"map('Product_2s', function (p) {
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:104]: should be Products

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 10]
```diff
@@ -3,101 +3,281 @@
...
+                                  { indexing: 'Search', storage: true, termVector: null }))
+                          }
+                     })"
+                };
+            }
+        }
+        #endregion
+        
+        #region dynamic_fields_7
+        public class Product_3
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:116]: separate namespace

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 11]
```diff
@@ -3,101 +3,281 @@
...
         #endregion
+        
+        #region dynamic_fields_8_JS
+        public class Products_ByProductType_JS : AbstractJavaScriptIndexCreationTask
+        {
+            public Products_ByProductType_JS()
+            {
+                Maps = new HashSet<string>
+                {
+                    @"map('Product_3s', function (p) {
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:157]: Products

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 12]
```diff
@@ -3,101 +3,281 @@
...
+                                { indexing: 'Search', storage: true, termVector: null })
+                        };
+                    })"
+                };
+            }
+        }
+        #endregion
+
+        #region dynamic_fields_10
+        public class Product_4
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:169]: definitely we need to use separate namespace, maybe separate files per each section?

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs [hunk 13]
```diff
@@ -3,101 +3,281 @@
...
+        #region dynamic_fields_11_JS
+        public class Attributes_ByName_JS : AbstractJavaScriptIndexCreationTask
         {
-            public CreateFieldItems_JavaScript()
+            public Attributes_ByName_JS()
             {
                 Maps = new HashSet<string>
                 {
-                    @"map('Products', function (p) {
+                    @"map('Product_4s', function (p) {
```
Comment (ppekrol) [inline Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs:221]: This is super confusing

Comment (Danielle9897) [reply Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs]: done:

* Renamed `Product_X` => `Product`  in all places & clients
* Used `Products` as the collection name

File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs (hunk 7 above)
Comment (Danielle9897) [reply Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/Indexes/DynamicFields.cs]: done: 

* using anonymous now
* removed class `IndexEntry`

**Review by ppekrol — APPROVED** [review]

**Pull request merged**